dbms/src/Interpreters/JoinUtils.h

```diff
@@ -71,13 +72,25 @@ struct ProbeProcessInfo
     size_t end_row;
     bool all_rows_joined_finish;
 
+    /// these are used for probe
+    bool prepare_for_probe_done = false;
+    Columns materialized_columns;
+    ColumnRawPtrs key_columns;
+    ColumnPtr null_map_holder = nullptr;
+    ConstNullMapPtr null_map = nullptr;
+    /// Used with ANY INNER JOIN
+    std::unique_ptr<IColumn::Filter> filter = nullptr;
+    /// Used with ALL ... JOIN
+    std::unique_ptr<IColumn::Offsets> offsets_to_replicate = nullptr;
+
     explicit ProbeProcessInfo(UInt64 max_block_size_)
         : max_block_size(max_block_size_)
         , min_result_block_size((max_block_size + 1) / 2)
         , all_rows_joined_finish(true){};
 
     void resetBlock(Block && block_, size_t partition_index_ = 0);
     void updateStartRow();
+    void prepareForProbe(const Names & key_names, const String & filter_column, ASTTableJoin::Kind kind, ASTTableJoin::Strictness strictness);
```

Comment:
It looks like this method doesn't work for cross joins, does it?

Reply:
Currently, cross join does not split the block during probe stage, so it is fine for now. I'm working on split cross join probe, and will update this method for cross join in that pr.

Reply:
ok